src/pycsou/opt/solver/lsmr.py

```diff
@@ -0,0 +1,332 @@
+
+    b: NDArray
+         (..., N) 'b' terms in the LSMR cost function. All problems are solved in parallel.
+    x0: NDArray
+        (..., N) initial point(s). Defaults to 0 if unspecified.
+    restart_rate: int
+        Number of iterations after which restart is applied. By default, restart is done after 'n' iterations, where 'n'
+        corresponds to the dimension of the linear operator :math:`\mathbf{A}`.
+
+    **Remark:** :py:class:`pycsou.opt.solver.stop.StopCriterion_LSMR` is developed for the stopping criterion of LSMR. For
+    computational speed, explicit norms were not computated. Instead, their estimation was used, which is referred from
+    [2]_.
+
+    Examples
+    --------
+    >>> import numpy as np
+     >>> from pycsou.abc import LinOp
+    >>> # Create a PSD linear operator
+    >>> rng = np.random.default_rng(seed=0)
+    >>> mat = rng.normal(size=(10, 10))
+    >>> A = LinOp.from_array(mat).gram()
+    >>> # Create the ground truth 'x_star'
+    >>> x_star = rng.normal(size=(2, 2, 10))
+    >>> # Generate the corresponding data vector 'b'
+    >>> b = A.apply(x_star)
+    >>> # Solve 'Ax=b' for 'x' with the conjugate gradient method
+    >>> lsmr = LSMR(A, show_progress=False)
+    >>> lsmr.fit(b=b)
+    >>> x_solution = lsmr.solution()
+    >>> assert np.allclose(x_star, x_solution)
+    True
+
+    **References:**
+
+    .. [2] D. C.-L. Fong and M. A. Saunders,
+        "LSMR: An iterative algorithm for sparse least-squares problems",
+        SIAM J. Sci. Comput., vol. 33, pp. 2950-2971, 2011.
+    """
+
+    def __init__(
+        self,
+        A: pyco.PosDefOp,
+        *,
+        damp: float = 0.0,
+        atol: float = 1e-06,
+        btol: float = 1e-06,
+        conlim: float = 1e08,
+        iter_lim: typ.Optional[int] = None,
+        folder: typ.Optional[pyct.PathLike] = None,
+        exist_ok: bool = False,
+        writeback_rate: typ.Optional[int] = None,
+        verbosity: int = 1,
+        show_progress: bool = True,
+        log_var: pyct.VarName = ("x",),
+    ):
+        super().__init__(
+            folder=folder,
+            exist_ok=exist_ok,
+            writeback_rate=writeback_rate,
+            verbosity=verbosity,
+            show_progress=show_progress,
+            log_var=log_var,
+        )
+
+        self._A = A
+        self._atol = atol
+        self._btol = btol
+        self._ctol = (1.0 / conlim) if (conlim > 0) else 0.0
+        self._iter_lim = (min(A.shape[0], A.shape[1])) if (iter_lim is None) else iter_lim
+        self._damp = damp
+        self._itn = 0
+        self._normb = None
+
+    @pycrt.enforce_precision(i=["b", "x0"], allow_None=True)
+    def m_init(
+        self,
+        b: pyct.NDArray,
+        x0: typ.Optional[pyct.NDArray] = None,
+    ):
+        xp = pycu.get_array_module(b)
+
+        b = xp.atleast_1d(b)
+        if b.ndim > 1:
+            b = b.squeeze()
+        m, n = self._A.shape
+
+        u = b
+        normb = xp.linalg.norm(b)
```

Comment:
Seems that beta is the norm in the dimension of interest (not the stacking dimensions). If this indeed the case, then  `normb = xp.linalg.norm(b, axis=-1, keepdims=True)` should do it.

Reply:
Fixed that.